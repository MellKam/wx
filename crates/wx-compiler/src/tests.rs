use super::*;
use indoc::indoc;
use insta::{assert_binary_snapshot, assert_yaml_snapshot};
use string_interner::{StringInterner, backend::StringBackend};

#[allow(unused)]
struct TestCase {
    files: files::Files,
    interner: StringInterner<StringBackend>,
    ast: ast::parser::ParserResult,
    hir: hir::builder::BuilderResult,
    mir: mir::MIR,
    module: wasm::Module,
    bytecode: Vec<u8>,
}

impl TestCase {
    fn assert_snapshot(name: &str, source: &str) {
        let mut interner = StringInterner::new();
        let mut files = files::Files::new();
        let file_id = files
            .add(name.to_string() + ".wx", source.to_string())
            .unwrap();

        let ast =
            ast::parser::Parser::parse(file_id, &files.get(file_id).unwrap().source, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_ast", ast);

        if ast.diagnostics.len() > 0 {
            return;
        }

        let hir = hir::Builder::build(&ast.ast, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_hir", hir);

        let mir = mir::Builder::build(&hir.hir);
        assert_yaml_snapshot!(name.to_string() + "_mir", mir);

        let module = wasm::Builder::build(&mir).unwrap();
        assert_yaml_snapshot!(name.to_string() + "_wasm", module);

        let bytecode = wasm::Encoder::encode(&module, &interner);
        assert_binary_snapshot!((name.to_string() + ".wasm").as_ref(), bytecode.clone());
    }
}

#[test]
fn factorial() {
    TestCase::assert_snapshot(
        "factorial",
        indoc! {"
        export func factorial(n: i32): i32 {
            if n <= 1 { return 1 };
            n * factorial(n - 1)
        }
    "},
    );
}

#[test]
fn pow() {
    TestCase::assert_snapshot(
        "pow",
        indoc! {"
        export func pow(base: i32, mut exp: i32): i32 {
            if exp < 0 {
                return 0; 
            };
            
            local mut result: i32 = 1;
            loop {
                if exp == 0 {
                break result; 
                };
                
                result *= base;
                exp -= 1;
            }
        }
    "},
    );
}

#[test]
fn func_pointers() {
    TestCase::assert_snapshot(
        "func_pointers",
        indoc! {"
        func add(a: i32, b: i32): i32 { a + b }
        func sub(a: i32, b: i32): i32 { a - b }

        func apply(
            binop: func(i32, i32) -> i32, 
            a: i32, 
            b: i32,
        ): i32 {
            binop(a, b)
        }

        export func main(): i32 {
            local a = apply(add, 5, 10);
            local b = apply(sub, 10, 5);
            local c: i32 = 5;

            a + b
        }
    "},
    );
}
