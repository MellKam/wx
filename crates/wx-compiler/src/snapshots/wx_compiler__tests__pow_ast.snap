---
source: crates/wx-compiler/src/tests.rs
expression: ast
---
ast:
  file_id: 0
  items:
    - kind:
        ExportModifier:
          export:
            start: 0
            end: 6
          item:
            kind:
              FunctionDefinition:
                signature:
                  name:
                    symbol: 1
                    span:
                      start: 12
                      end: 15
                  params:
                    open:
                      start: 15
                      end: 16
                    inner:
                      - inner:
                          mutable: ~
                          name:
                            symbol: 2
                            span:
                              start: 16
                              end: 20
                          annotation:
                            separator:
                              start: 20
                              end: 21
                            ty:
                              kind:
                                Identifier:
                                  symbol: 3
                              span:
                                start: 22
                                end: 25
                        span:
                          start: 16
                          end: 25
                        separator:
                          start: 25
                          end: 26
                      - inner:
                          mutable:
                            start: 27
                            end: 30
                          name:
                            symbol: 4
                            span:
                              start: 31
                              end: 34
                          annotation:
                            separator:
                              start: 34
                              end: 35
                            ty:
                              kind:
                                Identifier:
                                  symbol: 3
                              span:
                                start: 36
                                end: 39
                        span:
                          start: 31
                          end: 39
                        separator: ~
                    close:
                      start: 39
                      end: 40
                  result:
                    separator:
                      start: 40
                      end: 41
                    ty:
                      kind:
                        Identifier:
                          symbol: 3
                      span:
                        start: 42
                        end: 45
                block:
                  kind:
                    Block:
                      open:
                        start: 46
                        end: 47
                      inner:
                        - inner:
                            Expression:
                              expr:
                                kind:
                                  IfElse:
                                    condition:
                                      kind:
                                        Binary:
                                          left:
                                            kind:
                                              Identifier:
                                                symbol: 4
                                            span:
                                              start: 55
                                              end: 58
                                          operator:
                                            kind: Less
                                            span:
                                              start: 59
                                              end: 60
                                          right:
                                            kind:
                                              Int:
                                                value: 0
                                            span:
                                              start: 61
                                              end: 62
                                      span:
                                        start: 55
                                        end: 62
                                    then_block:
                                      kind:
                                        Block:
                                          open:
                                            start: 63
                                            end: 64
                                          inner:
                                            - inner:
                                                Expression:
                                                  expr:
                                                    kind:
                                                      Return:
                                                        value:
                                                          kind:
                                                            Int:
                                                              value: 0
                                                          span:
                                                            start: 80
                                                            end: 81
                                                    span:
                                                      start: 73
                                                      end: 81
                                              span:
                                                start: 73
                                                end: 81
                                              separator:
                                                start: 81
                                                end: 82
                                          close:
                                            start: 88
                                            end: 89
                                      span:
                                        start: 63
                                        end: 89
                                    else_block: ~
                                span:
                                  start: 52
                                  end: 89
                          span:
                            start: 52
                            end: 89
                          separator:
                            start: 89
                            end: 90
                        - inner:
                            LocalDefinition:
                              mutable:
                                start: 106
                                end: 109
                              name:
                                symbol: 5
                                span:
                                  start: 110
                                  end: 116
                              annotation:
                                separator:
                                  start: 116
                                  end: 117
                                ty:
                                  kind:
                                    Identifier:
                                      symbol: 3
                                  span:
                                    start: 118
                                    end: 121
                              eq:
                                start: 122
                                end: 123
                              value:
                                kind:
                                  Int:
                                    value: 1
                                span:
                                  start: 124
                                  end: 125
                          span:
                            start: 100
                            end: 125
                          separator:
                            start: 125
                            end: 126
                        - inner:
                            Expression:
                              expr:
                                kind:
                                  Loop:
                                    block:
                                      kind:
                                        Block:
                                          open:
                                            start: 136
                                            end: 137
                                          inner:
                                            - inner:
                                                Expression:
                                                  expr:
                                                    kind:
                                                      IfElse:
                                                        condition:
                                                          kind:
                                                            Binary:
                                                              left:
                                                                kind:
                                                                  Identifier:
                                                                    symbol: 4
                                                                span:
                                                                  start: 149
                                                                  end: 152
                                                              operator:
                                                                kind: Eq
                                                                span:
                                                                  start: 153
                                                                  end: 155
                                                              right:
                                                                kind:
                                                                  Int:
                                                                    value: 0
                                                                span:
                                                                  start: 156
                                                                  end: 157
                                                          span:
                                                            start: 149
                                                            end: 157
                                                        then_block:
                                                          kind:
                                                            Block:
                                                              open:
                                                                start: 158
                                                                end: 159
                                                              inner:
                                                                - inner:
                                                                    Expression:
                                                                      expr:
                                                                        kind:
                                                                          Break:
                                                                            label: ~
                                                                            value:
                                                                              kind:
                                                                                Identifier:
                                                                                  symbol: 5
                                                                              span:
                                                                                start: 178
                                                                                end: 184
                                                                        span:
                                                                          start: 172
                                                                          end: 184
                                                                  span:
                                                                    start: 172
                                                                    end: 184
                                                                  separator:
                                                                    start: 184
                                                                    end: 185
                                                              close:
                                                                start: 195
                                                                end: 196
                                                          span:
                                                            start: 158
                                                            end: 196
                                                        else_block: ~
                                                    span:
                                                      start: 146
                                                      end: 196
                                              span:
                                                start: 146
                                                end: 196
                                              separator:
                                                start: 196
                                                end: 197
                                            - inner:
                                                Expression:
                                                  expr:
                                                    kind:
                                                      Binary:
                                                        left:
                                                          kind:
                                                            Identifier:
                                                              symbol: 5
                                                          span:
                                                            start: 215
                                                            end: 221
                                                        operator:
                                                          kind: MulAssign
                                                          span:
                                                            start: 222
                                                            end: 224
                                                        right:
                                                          kind:
                                                            Identifier:
                                                              symbol: 2
                                                          span:
                                                            start: 225
                                                            end: 229
                                                    span:
                                                      start: 215
                                                      end: 229
                                              span:
                                                start: 215
                                                end: 229
                                              separator:
                                                start: 229
                                                end: 230
                                            - inner:
                                                Expression:
                                                  expr:
                                                    kind:
                                                      Binary:
                                                        left:
                                                          kind:
                                                            Identifier:
                                                              symbol: 4
                                                          span:
                                                            start: 239
                                                            end: 242
                                                        operator:
                                                          kind: SubAssign
                                                          span:
                                                            start: 243
                                                            end: 245
                                                        right:
                                                          kind:
                                                            Int:
                                                              value: 1
                                                          span:
                                                            start: 246
                                                            end: 247
                                                    span:
                                                      start: 239
                                                      end: 247
                                              span:
                                                start: 239
                                                end: 247
                                              separator:
                                                start: 247
                                                end: 248
                                          close:
                                            start: 253
                                            end: 254
                                      span:
                                        start: 136
                                        end: 254
                                span:
                                  start: 131
                                  end: 254
                          span:
                            start: 131
                            end: 254
                          separator: ~
                      close:
                        start: 255
                        end: 256
                  span:
                    start: 46
                    end: 256
            span:
              start: 7
              end: 256
      span:
        start: 0
        end: 256
diagnostics: []
